idxvpp2/idxvpp.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2017 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package idxvpp2
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/ligato/cn-infra/core"
+	"github.com/ligato/cn-infra/idxmap"
+	"github.com/ligato/cn-infra/idxmap/mem"
+	"github.com/ligato/cn-infra/logging"
+)
+
+// WithSwIfIndex is interface that items with sw_if_index must implement to get
+// indexed by SwIfIndex.
+type WithSwIfIndex interface {
+	// GetSwIfIndex should return sw_if_index assigned to the item.
+	GetSwIfIndex() uint32
+}
+
+// SwIfIndex is the "user API" to the registry of VPP items indexed by sw_if_index.
+// It provides read-only access intended for plugins that need to do the conversions
+// between logical names from NB and VPP IDs.
+type SwIfIndex interface {
+	// LookupByName retrieves a previously stored item identified by
+	// <name>. If there is no item associated with the give name in the mapping,
+	// the <exists> is returned as *false* and <item> as *nil*.
+	LookupByName(name string) (item WithSwIfIndex, exists bool)
+
+	// LookupIdx retrieves a previously stored item identified in VPP by the given
+	// <swIfIndex>.
+	// If there is no item associated with the given index, <exists> is returned
+	// as *false* with <name> and <item> both set to empty values.
+	LookupBySwIfIndex(swIfIndex uint32) (name string, item WithSwIfIndex, exists bool)
+
+	// WatchItems subscribes to receive notifications about the changes in the
+	// mapping related to items with sw_if_index.
+	WatchItems(subscriber core.PluginName, channel chan<- SwIfIndexDto)
```

Comment:
I think we should get rid of the coupling to plugins here and just use `string`.